File: Persistencia/ejercicios5/SistemaFarmacia.py
import pickle
import os

class Medicamento:
    def __init__(self, nombre="", cod_medicamento="", tipo="", precio=0.0):
        self.nombre = nombre
        self.cod_medicamento = cod_medicamento
        self.tipo = tipo
        self.precio = precio

    def leer(self):
        self.nombre = input("Nombre: ")
        self.cod_medicamento = input("Código: ")
        self.tipo = input("Tipo: ")
        self.precio = float(input("Precio: "))

    def mostrar(self):
        print(f"Nombre: {self.nombre}, Código: {self.cod_medicamento}, Tipo: {self.tipo}, Precio: {self.precio}")

class Farmacia:
    def __init__(self):
        self.nombre_farmacia = ""
        self.sucursal = 0
        self.direccion = ""
        self.medicamentos = []

    def leer(self):
        self.nombre_farmacia = input("Nombre Farmacia: ")
        self.sucursal = int(input("Sucursal: "))
        self.direccion = input("Dirección: ")
        nro_medicamentos = int(input("Nro medicamentos: "))
        self.medicamentos = []
        for i in range(nro_medicamentos):
            print(f"Medicamento {i+1}:")
            m = Medicamento()
            m.leer()
            self.medicamentos.append(m)

    def mostrar(self):
        print(f"\nFarmacia: {self.nombre_farmacia} | Sucursal: {self.sucursal} | Dirección: {self.direccion}")
        for m in self.medicamentos:
            m.mostrar()

    def contiene_medicamento(self, nombre):
        return any(m.nombre.lower() == nombre.lower() for m in self.medicamentos)

    def mostrar_medicamentos_por_tipo(self, tipo):
        for m in self.medicamentos:
            if m.tipo.lower() == tipo.lower():
                m.mostrar()

class ArchFarmacia:
    def __init__(self, archivo_nombre="farmacias.pkl"):
        self.archivo_nombre = archivo_nombre
        self.farmacias = []
        self._cargar_desde_archivo()

    def crear_archivos(self):
        self.farmacias = []
        self._guardar_en_archivo()

    def adicionar(self, farmacia):
        self.farmacias.append(farmacia)
        self._guardar_en_archivo()

    def listar(self):
        for f in self.farmacias:
            f.mostrar()

    def mostrar_medicamentos_tos(self, sucursal_x):
        for f in self.farmacias:
            if f.sucursal == sucursal_x:
                print(f"\n--- Medicamentos para la TOS en Sucursal {sucursal_x} ---")
                f.mostrar_medicamentos_por_tipo("tos")

    def mostrar_sucursales_con_medicamento(self, nombre_medicamento):
        print(f"\nFarmacias con el medicamento '{nombre_medicamento}':")
        for f in self.farmacias:
            if f.contiene_medicamento(nombre_medicamento):
                print(f"Sucursal: {f.sucursal} | Dirección: {f.direccion}")

    def _guardar_en_archivo(self):
        with open(self.archivo_nombre, "wb") as f:
            pickle.dump(self.farmacias, f)

    def _cargar_desde_archivo(self):
        if os.path.exists(self.archivo_nombre):
            with open(self.archivo_nombre, "rb") as f:
                self.farmacias = pickle.load(f)
        else:
            self.farmacias = []

# Programa principal para probar
def main():
    arch = ArchFarmacia()
    arch.crear_archivos()  # Comentar para no borrar datos anteriores

    for i in range(2):
        print(f"\n--- Ingresando Farmacia {i + 1} ---")
        f = Farmacia()
        f.leer()
        arch.adicionar(f)

    arch.listar()

    arch.mostrar_medicamentos_tos(1)
    arch.mostrar_sucursales_con_medicamento("Golpex")

if __name__ == "__main__":
    main()

File: Persistencia/ejercicios5/SistemaFarmacia.java
package Persistencia.ejercicios5;

import java.io.*;
import java.util.*;

public class SistemaFarmacia implements Serializable {

    // ---------- CLASE MEDICAMENTO ----------
    static class Medicamento implements Serializable {
        private String nombre;
        private String codMedicamento;
        private String tipo;
        private float precio;

        public void leer(Scanner sc) {
            System.out.print("Nombre: ");
            nombre = sc.nextLine();
            System.out.print("Código: ");
            codMedicamento = sc.nextLine();
            System.out.print("Tipo: ");
            tipo = sc.nextLine();
            System.out.print("Precio: ");
            precio = Float.parseFloat(sc.nextLine());
        }

        public void mostrar() {
            System.out.println("Nombre: " + nombre + ", Código: " + codMedicamento + ", Tipo: " + tipo + ", Precio: " + precio);
        }

        public String getTipo() {
            return tipo;
        }

        public float getPrecio() {
            return precio;
        }

        public String getNombre() {
            return nombre;
        }
    }

    // ---------- CLASE FARMACIA ----------
    static class Farmacia implements Serializable {
        private String nombreFarmacia;
        private int sucursal;
        private String direccion;
        private int nroMedicamentos;
        private Medicamento[] medicamentos = new Medicamento[100];

        public void leer(Scanner sc) {
            System.out.print("Nombre Farmacia: ");
            nombreFarmacia = sc.nextLine();
            System.out.print("Sucursal: ");
            sucursal = Integer.parseInt(sc.nextLine());
            System.out.print("Dirección: ");
            direccion = sc.nextLine();
            System.out.print("Nro medicamentos: ");
            nroMedicamentos = Integer.parseInt(sc.nextLine());

            for (int i = 0; i < nroMedicamentos; i++) {
                medicamentos[i] = new Medicamento();
                System.out.println("Medicamento " + (i + 1) + ":");
                medicamentos[i].leer(sc);
            }
        }

        public void mostrar() {
            System.out.println("\nFarmacia: " + nombreFarmacia + " | Sucursal: " + sucursal + " | Dirección: " + direccion);
            for (int i = 0; i < nroMedicamentos; i++) {
                medicamentos[i].mostrar();
            }
        }

        public String getDireccion() {
            return direccion;
        }

        public int getSucursal() {
            return sucursal;
        }

        public boolean contieneMedicamento(String nombre) {
            for (int i = 0; i < nroMedicamentos; i++) {
                if (medicamentos[i].getNombre().equalsIgnoreCase(nombre)) {
                    return true;
                }
            }
            return false;
        }

        public void mostrarMedicamentosPorTipo(String tipo) {
            for (int i = 0; i < nroMedicamentos; i++) {
                if (medicamentos[i].getTipo().equalsIgnoreCase(tipo)) {
                    medicamentos[i].mostrar();
                }
            }
        }
    }

    // ---------- CLASE ARCHIVO ----------
    static class ArchFarmacia implements Serializable {
        private ArrayList<Farmacia> farmacias = new ArrayList<>();
        private static final String ARCHIVO = "farmacias.dat";

        public void crearArchivos() {
            farmacias = new ArrayList<>();
            guardarEnArchivo();
        }

        public void adicionar(Farmacia f) {
            farmacias.add(f);
            guardarEnArchivo();
        }

        public void listar() {
            for (Farmacia f : farmacias) {
                f.mostrar();
            }
        }

        // b) Mostrar medicamentos tipo "tos" en sucursal específica
        public void mostrarMedicamentosTos(int sucursalX) {
            for (Farmacia f : farmacias) {
                if (f.getSucursal() == sucursalX) {
                    System.out.println("\n--- Medicamentos para la TOS en Sucursal " + sucursalX + " ---");
                    f.mostrarMedicamentosPorTipo("tos");
                }
            }
        }

        // c) Mostrar sucursales que tienen un medicamento específico
        public void mostrarSucursalesConMedicamento(String nombreMedicamento) {
            System.out.println("\nFarmacias con el medicamento '" + nombreMedicamento + "':");
            for (Farmacia f : farmacias) {
                if (f.contieneMedicamento(nombreMedicamento)) {
                    System.out.println("Sucursal: " + f.getSucursal() + " | Dirección: " + f.getDireccion());
                }
            }
        }

        public void guardarEnArchivo() {
            try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(ARCHIVO))) {
                oos.writeObject(farmacias);
            } catch (IOException e) {
                System.out.println("Error guardando archivo: " + e.getMessage());
            }
        }

        @SuppressWarnings("unchecked")
        public void cargarDesdeArchivo() {
            File f = new File(ARCHIVO);
            if (f.exists()) {
                try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(ARCHIVO))) {
                    farmacias = (ArrayList<Farmacia>) ois.readObject();
                } catch (IOException | ClassNotFoundException e) {
                    System.out.println("Error cargando archivo: " + e.getMessage());
                    farmacias = new ArrayList<>();
                }
            } else {
                farmacias = new ArrayList<>();
            }
        }
    }

    // ---------- MÉTODO PRINCIPAL ----------
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        ArchFarmacia arch = new ArchFarmacia();
        arch.cargarDesdeArchivo();

        System.out.println("¿Desea crear un nuevo archivo? (S/N)");
        String respuesta = sc.nextLine();
        if (respuesta.equalsIgnoreCase("S")) {
            arch.crearArchivos();
        }

        for (int i = 0; i < 2; i++) {
            System.out.println("\n--- Ingresando Farmacia " + (i + 1) + " ---");
            Farmacia f = new Farmacia();
            f.leer(sc);
            arch.adicionar(f);
        }

        System.out.println("\n--- Listado de Farmacias ---");
        arch.listar();

        // b) Mostrar medicamentos para la tos en sucursal 1
        arch.mostrarMedicamentosTos(1);

        // c) Mostrar sucursales que tienen "Golpex"
        arch.mostrarSucursalesConMedicamento("Golpex");
    }
}
